### ISSUE TITLE
[Feature]: Kernel Dispatch Overrides (in pursuit of deterministic execution)

### ISSUE THREAD

**Issue by bwasti** [vllm-project/vllm#25404]
### 🚀 The feature, motivation and pitch

I am working towards high-fidelity (+ bit-wise) numerical stability and would like overrides to be possible and easy to configure.  For reference, please see the proof of concept from Horace here: #24583 and associated blog post: https://thinkingmachines.ai/blog/defeating-nondeterminism-in-llm-inference/

## Per-kernel Overrides

The proposal here is to ease the vLLM framework in that direction by exposing low-level overrides for kernel dispatching.  Rather than a global "deterministic" override, these would be *per* kernel and ultimately live in configuration files that enumerate all overrides needed for a given service.

For example, https://github.com/vllm-project/vllm/blob/main/csrc/layernorm_kernels.cu#L416

```diff
  + auto deterministic = parse_env(KERN_OVERRIDE_FUSED_ADD_RMS_NORM_DETERMINISTIC)
  - if (ptrs_are_aligned && offsets_are_multiple_of_vector_width) {
  + if (ptrs_are_aligned && offsets_are_multiple_of_vector_width && !deterministic) {
    LAUNCH_FUSED_ADD_RMS_NORM(8);
  } else {
    LAUNCH_FUSED_ADD_RMS_NORM(0);
  }
```

By exposing configurable overrides (with a single layer of indirection in the form of configuration files), we allow users to explicitly make the tradeoffs between numerical stability and performance.  Some kernels, if tuned exclusively for determinism, are such a large hit to perf that it may not be worth it.  Others are extremely cheap and can fix large numerical gaps.  The framework should not be responsible for tuning across these, but instead provide sane defaults.

## Import `batch_invariant_ops`

We will either want to create a dependency on the project from thinking machines or maintain a similar one within vllm.
https://github.com/thinking-machines-lab/batch_invariant_ops/tree/main

The MIT License might not be compatible (not sure) with vLLMs Apache.  We may also want to override many more kernels.

## Configuration files

Configurations should simply enumerate the environment variables required to run the program.  e.g.

```bash
KERN_OVERRIDE_FUSED_ADD_RMS_NORM_DETERMINISTIC=1
KERN_OVERRIDE_FLEX_ATTN_DETERMINISTIC_SPLIT_TILE_SIZE=4096
KERN_OVERRIDE_IMPORT_BATCH_INVARIANT=1 # mm, addmm, softmax, mean
VLLM_ATTENTION_BACKEND=FLEX_ATTN
```

### Alternatives

Alternatively we can go 100% into a single deterministic configuration.  This would be quite slow, but would certainly be easier interface-wise.  I believe providing a sane default configuration file within vLLM can achieve the same ease-of-use while simultaneously exposing nuance for advanced users.

### Additional context

It is important to note that there hasn't been much explicit need for bit-wise accurate execution in the ML space, and recent development in RL are the first to show the importance of numerical precision and stability.  Because this is a very fast moving space it may not be worth over-investing in this.  We should keep change to vLLM to an absolute minimum to avoid churn and code stability risk.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ProExpertProg** [comment]
Is there a way to get torch.compile to generate deterministic kernels? I think that would be the simplest path to supporting this generally

**Comment by simon-mo** [comment]
We can (and should) bring in the kernels into vLLM with proper acknolwdegement so we can maintain and test them going forward (and distribute them)

**Comment by simon-mo** [comment]
Also I think a global deterministic mode is P0, then per op deterministic is more like a P2 on the priority level. Please ensure we can land this fast.

**Comment by Chillee** [comment]
@ProExpertProg There is some interest in that on the `torch.compile` side (cc: @eellison) but to some extent there is some interaction with vllm serving broadly (particularly attention).

Also, totally happy for folks to bring the kernels into vLLM - I think that's the path that makes sense as opposed to incurring the dependency.

**Comment by bwasti** [comment]
> torch.compile

cc @zou3519 as well

**Comment by WoosukKwon** [comment]
I don't understand the benefits of fine-grained configuration, compared to defining a global flag.
Also, manipulating kernel dispatching in vLLM won't be enough, as PyTorch itself (e.g., `torch.softmax`) is not batch-invariant in general.

**Comment by bwasti** [comment]
regarding the fine-grained vs single hard-coded flag, perhaps the example I gave was not very instructive since there aren't any tile-sizes to set.  consider instead how sglang does it for flashinfer:

```
get_int_env_var("SGLANG_FLASHINFER_PREFILL_SPLIT_TILE_SIZE", 4096)
```
^ this type of stuff will need centralization if we don't just want random hard-coded tiles sizes. there's a ton of ways to actually achieve determinism, its not a single kernel configuration. Further, per-kernel registration is basically free - just a single layer of indirection that'll forcibly document all the locations of determinism in a single file (which feels quite useful)

If folks are opposed to this, no problem. I'll just do whichever flags make sense as they come up

@WoosukKwon check out the linked batch_invariant_operators repo, you just need to use `torch.library.Library` to do this.  it can all happen within vllm

**Comment by simon-mo** [comment]
@bwasti please put up a PR first that demonstrate it works end to end.

**Comment by WoosukKwon** [comment]
@bwasti I see. I think what we want atm is a single flag (maybe env variable) that just works, while in the future we may want to config the tile sizes in details. I'd suggest making opinionated choices for now, and gradually make it more configurable.
